### PR TITLE
refactor: image manager

### PR DESCRIPTION
**Pull request by fuweid**
### Ⅰ. Describe what this PR did

Refactor image manager and make it stable.

Before this PR, we always meet the "not found" during `pouch image inspect`, `pouch rmi`. We have several PR to fix the problem. But it always show up.  In order to make the pouch stable, I decides to refactor. 

### Ⅱ. Does this pull request fix one issue?

NONE

### Ⅲ. Describe how you did it

I redesign the `imageStore` in ImageMgr. The design is here:

```
Primary reference is the reference used for pulling the image. For example,

 	pouch pull busybox:1.26

the busybox:1.26 is the primary reference. However, the user can use the following references to consume the busybox:1.26:

        busybox:1.26
        busybox@sha256:be3c11fdba7cfe299214e46edc642e09514dbb9bbefcd0d3836c05a1e0cd0642
        c30178c5239f

The Name[:Tag][@Digest] are the references generated from busybox:1.26. 
Therefore, We can call it searchable references, which allows users use different reference to use the busybox:1.26.

Searchable reference can be named by user. 
For example, the localhost:5000/busybox:latest is the searchable reference for the busybox:1.26. 
Then the localhost:5000/busybox:latest is like alias. You can remove the alias and the image is still here.

However, if we remove the primary reference, bot the image and all searchable references will be removed.

In the Containerd, the Name to be pulled is used as primary key. 
We can only use the name to search the image in Containerd. 
Based on this, we should make primary and searchable references concept here.

Besides this, the user can use busybox[:whatever]@digest to pull image. 
The registry doesn't care the tag information if you provide digest. 
The tag maybe wrong. In order to avoid this kind of data, we will trim tag information from Name:Tag@Digest reference.
```

### Ⅳ. Describe how to verify it

First, pull the `reg.docker.alibaba-inc.com/base/busybox:latest`
```
➜  ~ pouch pull reg.docker.alibaba-inc.com/base/busybox:latest
reg.docker.alibaba-inc.com/base/busybox:latest:                                   resolved       |++++++++++++++++++++++++++++++++++++++|
manifest-sha256:68effe31a4ae8312e47f54bec52d1fc925908009ce7e6f734e1b54a4169081c5: done           |++++++++++++++++++++++++++++++++++++++|
layer-sha256:04176c8b224aa0eb9942af765f66dae866f436e75acef028fe44b8a98e045515:    done           |++++++++++++++++++++++++++++++++++++++|
config-sha256:00f017a8c2a6e1fe2ffd05c281f27d069d2a99323a8cd514dd35f228ba26d2ff:   done           |++++++++++++++++++++++++++++++++++++++|
elapsed: 0.0 s
```

#### About inspect 
The following commands have the same result. 
```
pouch image inspect reg.docker.alibaba-inc.com/base/busybox:latest
pouch image inspect reg.docker.alibaba-inc.com/base/busybox@sha256:68effe31a4ae8312e47f54bec52d1fc925908009ce7e6f734e1b54a4169081c5
pouch image inspect reg.docker.alibaba-inc.com/base/busybox
pouch image inspect 00f017a8c2a6e1
```

#### About rmi

The following commands have the same result.
```
pouch rmi reg.docker.alibaba-inc.com/base/busybox:latest
pouch rmi reg.docker.alibaba-inc.com/base/busybox
pouch rmi 00f017a8c2a6e1
```

However, the `reg.docker.alibaba-inc.com/base/busybox@sha256:68effe31a4ae8312e47f54bec52d1fc925908009ce7e6f734e1b54a4169081c5` is the searchable reference. If we remove this, the image is still here because the primary reference is still here.

```
➜  ~ pouch rmi reg.docker.alibaba-inc.com/base/busybox@sha256:68effe31a4ae8312e47f54bec52d1fc925908009ce7e6f734e1b54a4169081c5
reg.docker.alibaba-inc.com/base/busybox@sha256:68effe31a4ae8312e47f54bec52d1fc925908009ce7e6f734e1b54a4169081c5
➜  ~ pouch images --digest
IMAGE ID       IMAGE NAME                                       DIGEST   SIZE
00f017a8c2a6   reg.docker.alibaba-inc.com/base/busybox:latest   <none>   686.62 KB
```

#### About pull

```
pouch pull busybox # equals => pouch pull `registry.hub.docker.com/library/busybox:latest`
pouch pull busybox:1.28 # equals => pouch pull `registry.hub.docker.com/library/busybox:1.228`
```

User can use `Name:Tag@Digest` to pull image.  However, the register doesn't care the tag information if the reference has the digest. In order to avoid the wrong tag information, we remove tag information in digest before pull.

Therefore, the following commands has the same result
```
pouch pull busybox:wrong@sha256:58ac43b2cc92c687a32c8be6278e50a063579655fe3090125dcb2af0ff9e1a64
pouch pull busybox@sha256:58ac43b2cc92c687a32c8be6278e50a063579655fe3090125dcb2af0ff9e1a64

➜  ~ pouch images
IMAGE ID       IMAGE NAME                                                                                                        SIZE
8ac48589692a   registry.hub.docker.com/library/busybox@sha256:58ac43b2cc92c687a32c8be6278e50a063579655fe3090125dcb2af0ff9e1a64   710.83 KB
```

### CODE REVIEW

**Comment by allencloud** [comment]
> Primary reference is the reference used for pulling the image. For example,
 	pouch pull busybox:1.26
the busybox:1.26 is the primary reference.

Is busybox:1.26 the only primary reference for the corresponding image? @fuweid

**Comment by fuweid** [comment]
HI, @allencloud sorry to reply late.

> Is busybox:1.26 the only primary reference for the corresponding image? @fuweid

No. Both `busybox@digest` and `busybox[:whatever]@digest` can be the primary reference for the corresponding image. One image has many primary references:

* Name:Tag
* Name@Digest
* Name:WrongTag@RightDigest [The Registry service doesn't care the tag information correct or not if the reference contains the digest]

In order to make the primary reference simple, the image, which comes from one registry service, has two primary references in pouchd:

* Name:Tag
* Name@Digest

BTW, the primary reference means that you use it to pull the image. :)

> And if we execute pouch pull busybox, in the command what is busybox?

If we execute `pouch pull busybox`, the pouchd will use `registry.hub.docker.com/library/busybox:latest` as the primary reference. 


Does it make senses to you?

File: daemon/mgr/image_store.go [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+	store.Lock()
+	defer store.Unlock()
+
+	// remove the relationship if the ref has been used by other
+	if oldP, ok := store.primaryRefIndexByRef[trimRef.String()]; ok {
+		// if the trimRef is primary reference
+		if oldP.String() == trimRef.String() {
+			if oldP.String() != trimPrimaryRef.String() {
+
+				return pkgerrors.Wrap(errtypes.ErrInvalidType, "invalid reference: cannot replace primary reference")
```
Comment (YaoZengzeng) [inline daemon/mgr/image_store.go]: The following command
```sh
pouch pull registry.hub.docker.com/library/busybox@sha256:552644f12d0cb92449213f701c88fabf8e5a7acfae3fcefd144b592293ded713

pouch pull busybox:1-uclibc
```
will make the pouchd report error like:
```sh
ERRO[2018-05-07T16:22:07.055029807+08:00] failed to pull image busybox:1-uclibc:1-uclibc: invalid reference: cannot replace primary reference: invalid type
```

From my point of view, if a Ref has already been a PrimaryRef, we may not need to maintain its relationship with other PrimaryRefs.

So maybe it is more reasonable that just `return nil` other then error here.

@fuweid WDYT?

Comment (fuweid) [reply daemon/mgr/image_store.go]: @YaoZengzeng , I think we should maintain the relationship.
If not, the searchable reference can override the primary relationship. In this case, we cannot clean the image in the containerd side. It maybe cause the dirty data in the host. WDYT?

I have updated my code. In this case, we should check the reference before we add it. Please take a look.

Thanks for the suggestion.

**Comment by YaoZengzeng** [comment]
LGTM

**Comment by allencloud** [comment]
Thanks for your review, @YaoZengzeng 
We are merging this.

**Pull request merged**